apollo-router/src/services/layers/query_analysis.rs

```diff
@@ -0,0 +1,150 @@
+use std::sync::Arc;
+
+use apollo_compiler::ApolloCompiler;
+use apollo_compiler::FileId;
+use apollo_compiler::HirDatabase;
+use http::StatusCode;
+use lru::LruCache;
+use tokio::sync::Mutex;
+
+use crate::services::SupergraphRequest;
+use crate::services::SupergraphResponse;
+use crate::spec::query::Operation;
+use crate::spec::Query;
+use crate::spec::Schema;
+use crate::Configuration;
+
+/// [`Layer`] for QueryParsing implementation.
+#[derive(Clone)]
+pub(crate) struct QueryAnalysisLayer {
+    /// set to None if QueryParsing is disabled
+    schema: Arc<Schema>,
+    configuration: Arc<Configuration>,
+    cache: Arc<Mutex<LruCache<QueryAnalysisKey, Arc<Mutex<ApolloCompiler>>>>>,
```

Comment:
this is a LRU cache without deduplication for now because the `DeduplicatingCache` API has strong requirements on the key and value (like serialization) that make it annoying to implement with a compiler. I'm planning to revamp the redis and cache API to make such changes easier